x/oracle/keeper/nonce.go

```diff
@@ -0,0 +1,175 @@
+			if n.FeederID == feederID {
+				nonce.NonceList = append(nonce.NonceList[:i], nonce.NonceList[i+1:]...)
+				if len(nonce.NonceList) == 0 {
+					k.removeNonceWithValidator(store, validator)
+				} else {
+					k.setNonce(store, nonce)
+				}
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// RemoveNonceWithFeederIDForValidators remove the nonce for a specific feederID from a set of validators
+func (k Keeper) RemoveNonceWithFeederIDForValidators(ctx sdk.Context, feederID uint64, validators []string) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.NonceKeyPrefix))
+	k.removeNonceWithFeederIDForValidators(store, feederID, validators)
+}
+
+// RemoveNonceWithFeederIDForAll remove the nonce for a specific feederID from all validators
+func (k Keeper) RemoveNonceWithFeederIDForAll(ctx sdk.Context, feederID uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.NonceKeyPrefix))
+	iterator := store.Iterator(nil, nil)
+	defer iterator.Close()
+	var validators []string
+	for ; iterator.Valid(); iterator.Next() {
+		var nonce types.ValidatorNonce
+		k.cdc.MustUnmarshal(iterator.Value(), &nonce)
+		validators = append(validators, nonce.Validator)
+	}
+	k.removeNonceWithFeederIDForValidators(store, feederID, validators)
+}
+
+// CheckAndIncreaseNonce check and increase the nonce for a specific validator and feederID
+func (k Keeper) CheckAndIncreaseNonce(ctx sdk.Context, validator string, feederID uint64, nonce uint32) (prevNonce uint32, err error) {
+	if nonce > uint32(common.MaxNonce) {
+		return 0, errors.New("nonce is too large")
+	}
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.NonceKeyPrefix))
+	if n, found := k.getNonce(store, validator); found {
+		for _, v := range n.NonceList {
+			if v.FeederID == feederID {
+				if v.Value+1 == nonce {
+					v.Value++
+					k.setNonce(store, n)
+					return nonce - 1, nil
+				}
+				return v.Value, errors.New("nonce is not consecutive")
+			}
+		}
+		return 0, errors.New("feeder not found")
+	}
+	return 0, errors.New("validator not found")
+}
+
+// internal usage for avoiding duplicated 'NewStore'
+
+func (k Keeper) getNonce(store prefix.Store, validator string) (types.ValidatorNonce, bool) {
+	bz := store.Get(types.NonceKey(validator))
+	if bz != nil {
+		var nonce types.ValidatorNonce
+		k.cdc.MustUnmarshal(bz, &nonce)
+		return nonce, true
+	}
+	return types.ValidatorNonce{}, false
+}
+
+func (k Keeper) setNonce(store prefix.Store, nonce types.ValidatorNonce) {
+	bz := k.cdc.MustMarshal(&nonce)
+	store.Set(types.NonceKey(nonce.Validator), bz)
+}
+
+func (k Keeper) removeNonceWithValidator(store prefix.Store, validator string) {
+	store.Delete(types.NonceKey(validator))
+}
+
+func (k Keeper) removeNonceWithValidatorAndFeederID(store prefix.Store, validator string, feederID uint64) bool {
+	if nonce, found := k.getNonce(store, validator); found {
+		for i, n := range nonce.NonceList {
+			if n.FeederID == feederID {
```

Comment:
Consider optimizing this loop by using some sort of a special key / value combination. For example we could store the key to be `validator` + `feederID`.